### PR TITLE
INF-548/mz/extend flyte shelltask

### PR DESCRIPTION
**Pull request by myz540**
# Shell Task extension
The updated `ShellTask` can be instantiated as follows:
```python
test_write_shell_task = ShellTask(
    name="test_write_shell_task",
    debug=True,
    script_file="/Users/mikezhong/test.sh",
    interpolizer=None,
    inputs=kwtypes(
        x=str, y=int, A=str, B=int, script_args=str, WORK_DIR=FlyteDirectory
    ),
    env=["A", "B", "WORK_DIR"],
    output_locs=[
        OutputLocation(
            var="k", var_type=FlyteFile, location="{inputs.WORK_DIR}/test_output.txt"
        )
    ],
)
```

You then use the shell task like:
```python
    test_file, test_dir = create_entities()
    x_int, y_str = t1(a=20)
    o = test_write_shell_task(
        x=y_str,
        y=x_int,
        A="This is A",
        B=99,
        script_args="arg_1 arg_2",
        WORK_DIR=test_dir,
    )
    test_read_shell_task(x=o)
    )

```

* The `interpolizer` will always be run on the `output_locs` for cases where the output directory or file need to be specified in the inputs. 
* `interpolizer`  will NOT run on the script if the `interpolizer=None` is used.
* The `env` param is a list of strings, this list should ALWAYS be a subset of `inputs`. These are pulled from the inputs, popped from `kwargs`, and set for the `ShellTask` execution
* A `script_args` keyword can be specified in the `inputs`. If detected, the bash script will be run with that string as the `$*`

Feels clunky and repetitive in the shell task definition. Not clear if `script_args` is being handled well

# TL;DR
Allows passing/setting environment variables for ShellTask
Allows passing arguments to ShellTask scripts

Outstanding issues:
* ShellTask execution occurs in cwd, for for local usage at least, any files written to static locations in the shell script will clobber or append on subsequent executions

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

### CODE REVIEW

File: flytekit/extras/tasks/shell.py [hunk 1]
```diff
@@ -109,6 +110,8 @@ def __init__(
         task_config: T = None,
         inputs: typing.Optional[typing.Dict[str, typing.Type]] = None,
         output_locs: typing.Optional[typing.List[OutputLocation]] = None,
+        env: typing.Optional[typing.Dict[str, typing.Any]] = None,
```
Comment (CalvinLeather) [inline flytekit/extras/tasks/shell.py]: If python were more expressive, it wouldn't actually be Any here right? Its like "classes whose __str__ are valid when interpolated into bash"? This is probably fine as is, as I think there just isn't a way to effectively communicate that (at least that I can think of?)

File: flytekit/extras/tasks/shell.py [hunk 2]
```diff
@@ -210,8 +217,22 @@ def execute(self, **kwargs) -> typing.Any:
             print(gen_script)
             print("\n==============================================\n")
 
+        if self._env is not None:
+            for k, v in self._env.items():
+                os.environ[k] = str(v)
```
Comment (CalvinLeather) [inline flytekit/extras/tasks/shell.py]: This has lasting side effects right? 

I googled because I wasn't sure, but subprocesss doesn't support a specific way of avoiding this. So, [recommendation](https://stackoverflow.com/questions/2231227/python-subprocess-popen-with-a-modified-environment) is to use os.environ.copy() and then return the environment to normal afterwards (I would write a quick test for this)

File: flytekit/extras/tasks/shell.py [hunk 3]
```diff
@@ -210,8 +217,22 @@ def execute(self, **kwargs) -> typing.Any:
...
+
+        working_dir = flytekit.current_context().working_directory
+        tmp_script_path = os.path.join(working_dir, "tmp_script.sh")
+        with open(tmp_script_path, 'w') as fp:
+            fp.write(gen_script)
+        os.chmod(tmp_script_path, stat.S_IRWXU)
+        if self._script_args:
```
Comment (CalvinLeather) [inline flytekit/extras/tasks/shell.py]: Oh neato

File: flytekit/extras/tasks/shell.py [hunk 4]
```diff
@@ -210,8 +217,22 @@ def execute(self, **kwargs) -> typing.Any:
             print(gen_script)
             print("\n==============================================\n")
 
+        if self._env is not None:
+            for k, v in self._env.items():
+                os.environ[k] = str(v)
+
+        working_dir = flytekit.current_context().working_directory
```
Comment (CalvinLeather) [inline flytekit/extras/tasks/shell.py:287]: https://github.com/flyteorg/flytekit/blob/386626356ff2cb810733a29a51cf3d6351c8424d/flytekit/core/utils.py#L170 has the source for this, looks like a pretty typically autodeleting temp dir, so look good to me. I think the only risk here is if the bash script being called for some reason looked for a tmp_script.sh (but unless someone passess in the temp directory path as an env variable, it seems unlikely they would know where to look for the tmp_script.sh anyway, so this seems extremely contrived)

File: flytekit/extras/tasks/shell.py [hunk 5]
```diff
@@ -210,8 +217,22 @@ def execute(self, **kwargs) -> typing.Any:
...
+            fp.write(gen_script)
+        os.chmod(tmp_script_path, stat.S_IRWXU)
+        if self._script_args:
+            full_script = tmp_script_path + " " + self._script_args
+        else:
+            full_script = tmp_script_path
+
         try:
-            subprocess.check_call(gen_script, shell=True)
+            subprocess.run(full_script, shell=True, check=True)
```
Comment (CalvinLeather) [inline flytekit/extras/tasks/shell.py:298]: Oof. bad job flight with no check=True. Thanks for fixing that!

**Review by CalvinLeather — APPROVED** [review]
Want to throw in a couple quick tests: one to ensure environment variable setting doesn't persist out of scope of the task, and another to ensure check=True functions as anticipated?

File: flytekit/extras/tasks/shell.py [hunk 6]
```diff
@@ -150,7 +155,9 @@ def __init__(
         self._script_file = script_file
         self._debug = debug
         self._output_locs = output_locs if output_locs else []
-        self._interpolizer = _PythonFStringInterpolizer()
+        self._interpolizer = interpolizer
+        self.__output_interpolizer = _PythonFStringInterpolizer()
```
Comment (myz540) [inline flytekit/extras/tasks/shell.py:209]: really ugly, I did end up needing to use the interpolizer on the output for instances where a directory or file name need to be passed in (for work to be written to, for instance), and then captured as output. Here, we need to interpolate the `input.x` into the output location

File: flytekit/extras/tasks/shell.py [hunk 7]
```diff
@@ -192,26 +199,54 @@ def execute(self, **kwargs) -> typing.Any:
         Executes the given script by substituting the inputs and outputs and extracts the outputs from the filesystem
         """
         logger.info(f"Running shell script as type {self.task_type}")
+        breakpoint()
+
+        # somewhat hidden behavior and hacky at the moment, if script_args is in kwargs, grab it and save it
+        script_args = kwargs.pop("script_args") if "script_args" in kwargs else None
```
Comment (myz540) [inline flytekit/extras/tasks/shell.py]: feels hacky

File: flytekit/extras/tasks/shell.py (hunk 5 above)
Comment (myz540) [reply flytekit/extras/tasks/shell.py:298]: I think `check_call()` is an older method that has `check=True` by default. I beleive all the various subprocess methods were consolidated and `run()` is now the recommended method with all the other methods collapsed into params

File: flytekit/extras/tasks/shell.py [hunk 8]
```diff
@@ -109,6 +110,8 @@ def __init__(
         task_config: T = None,
         inputs: typing.Optional[typing.Dict[str, typing.Type]] = None,
         output_locs: typing.Optional[typing.List[OutputLocation]] = None,
+        env: typing.Optional[typing.List[str]] = None,
```
Comment (CalvinLeather) [inline flytekit/extras/tasks/shell.py]: Will BOLO below, I would have expected this to be `typing.Optional[typing.Dict[str, typing.Type]]`

Comment (CalvinLeather) [reply flytekit/extras/tasks/shell.py]: (i.e., because env variables are named)

File: flytekit/extras/tasks/shell.py (hunk 7 above)
Comment (CalvinLeather) [reply flytekit/extras/tasks/shell.py]: Here is a less hacky way of doing this (I think)
Instead of passing all kwargs to interpolate (which is why you need to pop here), use a list comp to grab the kwargs that match the specification defined in the inputs variable passed to __init__. Then you can leave anything in kwargs, and even just add script_args to the function definition so it isn't included in kwargs (probably)... does that make sense?

Comment (CalvinLeather) [reply flytekit/extras/tasks/shell.py]: Ditto for output_locs

File: flytekit/extras/tasks/shell.py [hunk 9]
```diff
@@ -192,26 +199,54 @@ def execute(self, **kwargs) -> typing.Any:
...
         if self._output_locs:
             for v in self._output_locs:
-                outputs[v.var] = self._interpolizer.interpolate(v.location, inputs=kwargs)
+                outputs[v.var] = self.__output_interpolizer.interpolate(v.location, inputs=kwargs)
+
+        # remove the extra ENV vars from kwargs and set them for the shell script
+        original_env = os.environ.copy()
+        if self._env:
+            for k in self._env:
+                if k in kwargs:
```
Comment (CalvinLeather) [inline flytekit/extras/tasks/shell.py]: Oh I see, you as passing them in as seperate kwargs. I was thinking you would just pass a dict in as one of the kwargs for execute (e.g., env = {} gets passed when calling the callable). That explains the typing above

File: flytekit/extras/tasks/shell.py (hunk 8 above)
Comment (CalvinLeather) [reply flytekit/extras/tasks/shell.py]: Ah okay I see, this is just the list of names of variables, which are then extracted from kwargs

File: flytekit/extras/tasks/shell.py (hunk 9 above)
Comment (CalvinLeather) [reply flytekit/extras/tasks/shell.py]: Not sure which is better... on one hand, if you pass a dict you need to either use a dynamic workflow to assemble the dict, or use a task to assemble it (i.e., you can't have a dict of flyte objects I think, the entire dict needs to be a flyte object/promise), however, this is probably harder to read because the env variables are mixed into the kwargs of the callable (so unless you prefix them all like "ENV_WHATEVER" so you can distinguish them, its harder to read)

File: flytekit/extras/tasks/shell.py (hunk 7 above)
Comment (myz540) [reply flytekit/extras/tasks/shell.py]: which function definition? You mean `execute()`? We cannot add any args to execute since we do not call it, we can only add key args that we specify in `inputs`. Basically, all variables in `inputs` are put into hte kwargs dict regardless of what we intend to use them for

File: flytekit/extras/tasks/shell.py (hunk 8 above)
Comment (myz540) [reply flytekit/extras/tasks/shell.py]: ya, adding additional typing info here felt redundant and opened room for mismatch

File: flytekit/extras/tasks/shell.py (hunk 9 above)
Comment (myz540) [reply flytekit/extras/tasks/shell.py]: My thinking was "Convention: All environment variables are UPPERCASE"

Comment (myz540) [reply flytekit/extras/tasks/shell.py]: passing a dict to the `inputs` could probably work but we can't restrict the set of keys at compile time, so any env vars could be added and there wouldn't be control over it. I'm not sure how complex inputs are handled on the flyte console UI when specifying params at launch

File: flytekit/extras/tasks/shell.py [hunk 10]
```diff
@@ -97,6 +98,35 @@ def interpolate(
 T = typing.TypeVar("T")
 
 
+class MappingRevisor:
```
Comment (CalvinLeather) [inline flytekit/extras/tasks/shell.py:101]: Docstring please describing the example usages of this, and why it is needed for os.environ?

File: flytekit/extras/tasks/shell.py [hunk 11]
```diff
@@ -97,6 +98,35 @@ def interpolate(
...
+       for k, v in self.changes.items():
+           self.mapping[k] = v
+       for k in self.new_keys:
+           self.mapping.pop(k)
+
+   def __enter__(self):
+       return self
+
+   def __exit__(self, *args, **kwargs):
+       self.reset()
```
Comment (CalvinLeather) [inline flytekit/extras/tasks/shell.py]: if you think reset is idempotent enough (I think it is?) I guess the `__del__` wouldn't be bad, and buttresses against misuse (i.e., someone who doesn't use this as a context manager)?

Comment (myz540) [reply flytekit/extras/tasks/shell.py]: ya, that's what I was thinking. `reset()` should be idempotent. Will add del now
